### PR TITLE
Add `graphql/getCurrentUserCodySubscription` endpoint

### PR DESCRIPTION
**Pull request by mkondratek**
- ~Needs https://github.com/sourcegraph/sourcegraph/pull/59742.~
- Needs https://github.com/sourcegraph/sourcegraph/pull/59816.
- Required for https://github.com/sourcegraph/jetbrains/issues/276.

---

> the schema changes should go live when dotcom gets deployed during these times: `09:00 UTC on Monday, 06:00 UTC rest of weekdays`

## Test plan
- [x] let's wait for tomorrow so that the schema changes go live and we can actually test that endpoint from the client
- [x] tested with the jetbrains client + I added a test case - works 🚀

### CODE REVIEW

File: lib/shared/src/sourcegraph-api/graphql/client.ts [hunk 1]
```diff
@@ -95,6 +95,18 @@ interface CurrentUserCodyProEnabledResponse {
     } | null
 }
 
+interface CurrentUserCodySubscriptionResponse {
+    currentUser: {
+        codySubscription: {
+            status: string
+            isPro: boolean
+            applyProRateLimits: string
+            currentPeriodStartAt: string
```
Comment (mkondratek) [inline lib/shared/src/sourcegraph-api/graphql/client.ts:104]: or should it be `Date`?

File: lib/shared/src/sourcegraph-api/graphql/client.ts [hunk 2]
```diff
@@ -170,6 +182,15 @@ export interface CodyLLMSiteConfiguration {
     provider?: string
 }
 
+
+export interface CurrentUserCodySubscription {
+    status: string
+    isPro: boolean
+    applyProRateLimits: string
+    currentPeriodStartAt: string
```
Comment (mkondratek) [inline lib/shared/src/sourcegraph-api/graphql/client.ts:190]: (`Date` here too?)

File: lib/shared/src/sourcegraph-api/graphql/client.ts [hunk 3]
```diff
@@ -95,6 +95,18 @@ interface CurrentUserCodyProEnabledResponse {
     } | null
 }
 
+interface CurrentUserCodySubscriptionResponse {
+    currentUser: {
+        codySubscription: {
+            status: string
+            isPro: boolean
```
Comment (kalanchan) [inline lib/shared/src/sourcegraph-api/graphql/client.ts:102]: the `codySubscription` schema has changed a bit, `isPro` is now `pro`

https://github.com/sourcegraph/sourcegraph/blob/da169dc09733d36a917e8bf96b79618a708a926a/cmd/frontend/graphqlbackend/schema.graphql#L6388

File: lib/shared/src/sourcegraph-api/graphql/client.ts (hunk 2 above)
Comment (kalanchan) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:190]: yeah should be datetime

https://github.com/sourcegraph/sourcegraph/blob/da169dc09733d36a917e8bf96b79618a708a926a/cmd/frontend/graphqlbackend/schema.graphql#L6388

File: lib/shared/src/sourcegraph-api/graphql/client.ts (hunk 1 above)
Comment (kalanchan) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:104]: yup should be datetime

https://github.com/sourcegraph/sourcegraph/blob/da169dc09733d36a917e8bf96b79618a708a926a/cmd/frontend/graphqlbackend/schema.graphql#L6388

**Review by kalanchan — CHANGES_REQUESTED** [review]

File: lib/shared/src/sourcegraph-api/graphql/client.ts (hunk 3 above)
Comment (mkondratek) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:102]: you mean the field is `plan` now, right?

Comment (kalanchan) [reply lib/shared/src/sourcegraph-api/graphql/client.ts]: woops yes good catch, sorry haha!

File: lib/shared/src/experimentation/FeatureFlagProvider.ts [hunk 4]
```diff
@@ -30,6 +30,13 @@ export enum FeatureFlag {
     // Enable Cody PLG features on JetBrains
     CodyProJetBrains = 'cody-pro-jetbrains',
 
+    // use-ssc-for-cody-subscription is a feature flag that enables the use of SSC as the source of truth for Cody subscription data.
+    UseSscForCodySubscription = 'use-ssc-for-cody-subscription',
+
+    // cody-pro-trial-ended is a feature flag that indicates if the Cody Pro "Free Trial"  has ended.
+    // (Enabling users to use Cody Pro for free for 3-months starting in late Q4'2023.)
+    CodyProTrialEnded = 'cody-pro-trial-ended',
+
```
Comment (mkondratek) [inline lib/shared/src/experimentation/FeatureFlagProvider.ts:39]: Note: I added these feature flags too.

**Review by vdavid — APPROVED** [review]
Looks pretty boilerplate-y, not much to go wrong here 👍

File: lib/shared/src/sourcegraph-api/graphql/client.ts [hunk 5]
```diff
@@ -95,6 +96,18 @@ interface CurrentUserCodyProEnabledResponse {
     } | null
 }
 
+interface CurrentUserCodySubscriptionResponse {
+    currentUser: {
```
Comment (pkukielka) [inline lib/shared/src/sourcegraph-api/graphql/client.ts:100]: Can you give me some examples what could be values of those fields?

File: agent/src/agent.ts [hunk 6]
```diff
@@ -542,6 +542,15 @@ export class Agent extends MessageHandler {
             return res.codyProEnabled
         })
 
+        this.registerRequest('graphql/getCurrentUserCodySubscription', async () => {
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply agent/src/agent.ts:545]: looks like a good point - @vdavid  , can you confirm?

File: lib/shared/src/sourcegraph-api/graphql/client.ts (hunk 5 above)
Comment (mkondratek) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:100]: please take a look at the related document:

https://docs.google.com/document/d/1p8V_c5wkmNgnCshRdlnjbHqxgkYLVlvpTyDdbHu4VSY/edit

File: agent/src/agent.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (vdavid) [reply agent/src/agent.ts:545]: Errm, could you rephrase the question? I've never used either function, so not sure what it does. I suppose `registerAuthenticatedRequest` only works if the user has a valid access token with dotcom, while `registerRequest` works without it? If this is the case, then I suppose we need the authed one because anonymous users don't have subscriptions.

[the remark this answers is not in the document]
Comment (mkondratek) [reply agent/src/agent.ts:545]:
```
    // Alternative to `registerRequest` that awaits on authentication changes to
    // propagate before calling the method handler.
    public registerAuthenticatedRequest<M extends RequestMethodName>(
...
```

It sounds correct to use `registerAuthenticatedRequest`

File: lib/shared/src/sourcegraph-api/graphql/client.ts (hunk 5 above)
Comment (pkukielka) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:100]: I was wondering about `applyProRateLimits` because name sounds like boolean but it is a string.
And it's not mentioned in the doc.

Comment (pkukielka) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:100]: Actually I see it's laso not in the `CURRENT_USER_CODY_SUBSCRIPTION_QUERY`

Comment (mkondratek) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:100]: good point! it's boolean

Comment (mkondratek) [reply lib/shared/src/sourcegraph-api/graphql/client.ts:100]: there is more context there: https://github.com/sourcegraph/sourcegraph/pull/59816/files

**Review by pkukielka — APPROVED** [review]
LGTM
